gwcs/wcs.py

```diff
@@ -634,3 +641,211 @@ def fix_inputs(self, fixed):
+            _store_2D_coefficients(hdr, fit_inv_poly_v, 'BP', keeplinear=True)
+            hdr['sipmxerr'] = (max_resid, 'Maximum difference from the GWCS model SIP is fit to.')
+            hdr['sipiverr'] = (max_inv_resid, 'Maximum difference for the inverse transform')
+
+        else:
+            hdr['ctype1'] = 'RA---TAN'
+            hdr['ctype2'] = 'DEC--TAN'
+
+        hdr['cd1_1'] = cd[0][0]
+        hdr['cd1_2'] = cd[0][1]
+        hdr['cd2_1'] = cd[1][0]
+        hdr['cd2_2'] = cd[1][1]
+        return hdr
+
+
+def _fit_2D_poly(degree, max_error, u, v, x, y, verbose=False):
+    """
+    Fit a pair of ordinary 2D polynomial to the supplied inputs (u, v) and
+    outputs (x, y)
+    """
+    llsqfitter = LinearLSQFitter()
+    # The case of one pass with the specified polynomial degree
+    if degree is None:
+        deglist = range(1,10)
+    else:
+        deglist = [degree]
+    for deg in deglist:
+        poly_x = Polynomial2D(degree=deg)
+        poly_y = Polynomial2D(degree=deg)
+        fit_poly_x = llsqfitter(poly_x, u, v, x)
+        fit_poly_y = llsqfitter(poly_y, u, v, y)
+        max_resid = _compute_distance_residual(x, y,
+                                               fit_poly_x(u, v), fit_poly_y(u, v))
+        if verbose:
+            print(f'Degree = {degree}, max_resid = {max_resid}')
+        if max_error is not None and max_resid < max_error:
+            if verbose:
+                print('terminating condition met')
+            break
+        if degree is None and deg == 9:
+            raise RuntimeError("Unable to achieve required fit accuracy with SIP polynomials")
+    return fit_poly_x, fit_poly_y, max_resid
+
+
+def _compute_distance_residual(undist_x, undist_y, fit_poly_x, fit_poly_y):
+    """
+    Compute the distance residuals and return the rms and maximum values.
+    """
+    dist = np.sqrt((undist_x - fit_poly_x)**2 + (undist_y - fit_poly_y)**2)
+    max_resid = dist.max()
+    return max_resid
+
+
+def _reform_poly_coefficients(fit_poly_x, fit_poly_y):
+    """
+    The fit polynomials must be recombined to align with the SIP decomposition
+
+    The result is the f(u,v) and g(u,v) polynomials, and the CD matrix.
+    """
+    # Extract values for CD matrix and recombining
+    c11 = fit_poly_x.c1_0.value
+    c12 = fit_poly_x.c0_1.value
+    c21 = fit_poly_y.c1_0.value
+    c22 = fit_poly_y.c0_1.value
+    sip_poly_x = fit_poly_x.copy()
+    sip_poly_y = fit_poly_y.copy()
+    # Force low order coefficients to be 0 as defined in SIP
+    sip_poly_x.c0_0 = 0
+    sip_poly_y.c0_0 = 0
+    sip_poly_x.c1_0 = 0
+    sip_poly_x.c0_1 = 0
+    sip_poly_y.c1_0 = 0
+    sip_poly_y.c0_1 = 0
+
+    cdmat = ((c11, c12), (c21, c22))
+    invcdmat = npla.inv(np.array(cdmat))
+    degree = fit_poly_x.degree
+    # Now loop through all remaining coefficients
+    for i in range(0, degree + 1):
+        for j in range(0, degree + 1):
+            if (i + j > 1) and (i + j < degree + 1):
```

Comment:
Even simple and even more efficient:
```suggestion
    for i in range(degree + 1):
        for j in range(max(0, 2 - i), degree + 1 - i):
```